packages/sdk-drop-in/src/lib/MoniteIframeAppCommunicator.ts

```diff
@@ -0,0 +1,221 @@
+export class MoniteIframeAppCommunicator {
+  private readonly targetWindow?: Window;
+  private readonly isHost: boolean;
+  private readonly listeners = new Map<
+    string,
+    (payload: ListenerPayload) => void
+  >();
+  private readonly slots = new Map<
+    string,
+    { payload: PayloadSlot; abortController?: AbortController }
+  >();
+
+  private connectionToHostRetryInterval?: number;
+  private slotQueue = new Set<string>();
+  private connected = false;
+
+  static allowedSlots = ['locale', 'theme', 'fetch-token'];
+
+  constructor(iframeElement: HTMLIFrameElement | Window) {
+    if (iframeElement instanceof HTMLIFrameElement) {
+      this.isHost = true;
+      this.targetWindow = iframeElement.contentWindow ?? undefined;
+    } else {
+      this.isHost = false;
+      this.targetWindow = iframeElement;
+    }
+  }
+
+  public connect() {
+    if (typeof window === 'undefined' || typeof document === 'undefined')
+      return void console.warn('Cannot connect in non-browser environment');
+
+    this.disconnect();
+
+    if (!this.targetWindow) throw new Error('Target window is not set');
+
+    window.addEventListener('message', this.onWindowMessage);
+
+    if (!this.isHost) {
+      this.connectionToHostRetryInterval = setInterval(() => {
+        if (!this.targetWindow) throw new Error('Target window is not set');
+        this.connect();
+      }, 200) as unknown as number;
+
+      this.targetWindow?.postMessage({ type: 'monite:init_host' }, '*');
+    }
+  }
+
+  private onWindowMessage = (event: MessageEvent) => {
+    if (!event.source || !('postMessage' in event.source))
+      throw new Error('Invalid source');
+
+    const data = event.data;
+
+    assertData(data);
+
+    if (data.type === 'monite:init_host') {
+      this.targetWindow?.postMessage({ type: 'monite:init_iframe' }, '*');
+      this.connected = true;
+    } else if (data.type === 'monite:init_iframe') {
+      this.connected = true;
+      clearInterval(this.connectionToHostRetryInterval);
+      this.runSlotQueue();
+    }
+
+    if (this.slots.has(data.type)) {
+      this.queueSlot(data.type);
+    }
+
+    if (MoniteIframeAppCommunicator.allowedSlots.includes(data.type))
+      this.listeners.get(data.type)?.(data.payload);
+
+    function assertData(
+      data: unknown
+    ): asserts data is { type: string; payload?: ListenerPayload } {
+      if (typeof data !== 'object' || data === null) {
+        throw new Error(`Invalid message data: ${data}`);
+      }
+      if (typeof (data as { type: unknown }).type !== 'string') {
```

Comment:
Should we handle a cases when other extentions are enabled and iterupt this payload? 

<img width="1202" alt="Screenshot 2024-07-02 at 17 51 57" src="https://github.com/team-monite/monite-sdk/assets/51091179/f0643253-4411-4034-8170-ccebcb151d60">

Reply:
Yep, we have to think about that too. We will find a way on how to do this. Probably, it's possible to check the origin of the event source, but we have to compare the origin with the customer domain, which is not easy (but doable).

Reply:
Btw, released to your comment: I refactored `console.log` usage to avoid displaying sensitive data.